cocoa/src/toga_cocoa/app.py

```diff
@@ -109,6 +109,7 @@ def __init__(self, interface):
 
         # Create the lookup table for commands and menu items
         self._menu_items = {}
+        self._active_window_before_dialog = None
```

Comment:
Quoting from my previous review: 

> ... on macOS, adding the "pre-display current window" as something that is tracked on app-modal dialogs seems like it should be reasonably straightforward **(and something that can be stored on the dialog)** 

(emphasis added)

Storing the state on the app means there's a circular dependency on setting the variable as soon as you have a one window with a dialog.

Reply:
I think I didn't understand what you meant. Did you meant to store the state on the native dialog object(i.e, on the NSAlert object itself?) or on the `dialog._impl`? When we get the key window in presence of a dialog with `NSApplication.keyWindow`, it returns the associated private `NSWindow` instance(i.e., `_NSAlertPanel`). So, how do we get to the `_impl` or to the `NSAlert` object from a `_NSAlertPanel`?

Also, I now realize that storing the previous active window may not be the correct thing to do here. For example:
* There are 2 windows open(`Window A`, `Window B`) and `Window A` has the current focus. Now, an `app modal dialog` opens and it stores `Window A` as the previously active window. 
* Since, the dialog is app modal, the user can manually change focus to other window, so the user manually changes focus to `window B` and then again changes focus on to the `app modal dialog`. 
* Now, if the app were to query the current window then it would return the cached `Window A` as the previously active window, when in reality `Window C` was the previously active window. Hence, giving incorrect results.

This is not a problem on winforms, as only 1 dialog can be shown at any given time. So, focus can't be changed to other windows of the app(Until the dialog is not dismissed). But on winforms, there is no way of knowing whether the focus is on a dialog or on something outside the app (`Form.ActiveForm` returns `None` in both cases). And even if the focus cannot be changed to other windows of the app(while the dialog is not dismissed), but the focus can be changed to other apps. So, returning the cached previously active window would be incorrect.

The more I look at the problem and the solution that this PR implements, the more I realize that I am going against the native behavior of the platform. I am trying to bend something that is not meant to be bent in the way I want. And the result is fragile and complicated.

Maybe the correct solution is to just prevent the error on macOS, leave the other platforms with their native behavior and just document the native platform quirks.

Reply:
> I think I didn't understand what you meant. Did you meant to store the state on the native dialog object(i.e, on the NSAlert object itself?) or on the `dialog._impl`? 

Either - just *not* the app.
 
> Also, I now realize that storing the previous active window may not be the correct thing to do here. For example:

Yes - this is exactly why I was flagging the state variable on the app as a problem - dialogs are separate from the app, and there's complex interactions at play here.

> The more I look at the problem and the solution that this PR implements, the more I realize that I am going against the native behavior of the platform. I am trying to bend something that is not meant to be bent in the way I want. And the result is fragile and complicated.
> 
> Maybe the correct solution is to just prevent the error on macOS, leave the other platforms with their native behavior and just document the native platform quirks.

... which is what I was suggesting before we went down this rabbit hole :-) The most recent iterations here are adding a lot of complex logic to handle a specific edge case that doesn't really have a clear and obvious answer. If a dialog is showing, the "current window" *is* a vague concept. 

Not crashing is the primary requirement. Beyond that, I don't see any problem with returning as much of an answer as any given platform will allow, which means that there may not be a current window on some platforms when a dialog is showing - and then documenting that as a platform quirk.